src/dmd/cond.d

```diff
@@ -362,13 +373,66 @@ extern (C++) final class StaticForeach : RootObject
         auto catass = new CatAssignExp(aloc, new IdentifierExp(aloc, idres), res[1]);
         s2.push(createForeach(aloc, pparams[1], new ExpStatement(aloc, catass)));
         s2.push(new ReturnStatement(aloc, new IdentifierExp(aloc, idres)));
-        auto aggr = wrapAndCall(aloc, new CompoundStatement(aloc, s2));
-        sc = sc.startCTFE();
-        aggr = aggr.expressionSemantic(sc);
-        aggr = resolveProperties(sc, aggr);
-        sc = sc.endCTFE();
-        aggr = aggr.optimize(WANTvalue);
-        aggr = aggr.ctfeInterpret();
+
+        Expression aggr;
+        Type indexty;
+
+        if (rangefe && (indexty = ety.typeSemantic(aloc, sc)).isintegral())
+        {
+            // min = (upr - lwr)
+            Expression min = new MinExp(aloc, rangefe.upr, rangefe.lwr);
+            min = min.expressionSemantic(sc);
+            min = min.optimize(WANTvalue);
+
+            // gt = (upr > lwr)
+            Expression gt = new CmpExp(TOK.greaterThan, aloc, rangefe.upr, rangefe.lwr);
+            gt = gt.expressionSemantic(sc);
+            gt = gt.optimize(WANTvalue);
+
+            // gtsz = (min > size_t)
+            auto szmax = new IntegerExp(aloc, size_t.max, Type.tsize_t);
+            Expression gtsz = new CmpExp(TOK.greaterThan, aloc, min, szmax);
+            gtsz = gtsz.expressionSemantic(sc);
+            gtsz = gtsz.optimize(WANTvalue);
+
+            size_t length;
+
+            if (gt.isBool(true))
+            {
+                // This can happen on 32bit DMD
+                if (gtsz.isBool(true))
+                    error(aloc, "`static %s` iterations cannot exceed `size_t.max`, range expression is `(%s .. %s)`", Token.toChars(rangefe.op), rangefe.lwr.toChars, rangefe.upr.toChars);
+                else
+                    length = cast(size_t) min.toInteger();
+            }
+
+            const lwr = rangefe.lwr.toInteger();
+            auto  upr = rangefe.upr.toInteger();
```

Comment:
```suggestion
            auto lwrRange = getIntRange(rangefe.lwr);
            auto uprRange = getIntRange(rangefe.upr);

            const lwr = rangefe.lwr.toInteger();
            auto  upr = rangefe.upr.toInteger();
            size_t length = 0;

            if (lwrRange.imin <= uprRange.imax)
            {
                auto bounds = IntRange.fromType(Type.tsize_t);
                if (!bounds.contains(uprRange - lwrRange))
                    error(aloc, "`static %s` iterations cannot exceed `size_t.max`, range expression is `(%s .. %s)`", Token.toChars(rangefe.op), rangefe.lwr.toChars, rangefe.upr.toChars);
                else
                    length = cast(size_t) upr - lwr;
            }

```

Reply:
You'll need to import both dmd.dcast and dmd.intrange.

Reply:
Thanks for the code.

I also tested IntRange and SignExtendedNumber they actually showed correct results but not what I need to replicate the current behaviour. Right now Foreach sema. gets the index type for the For statement using this code:
```
scope AddExp ea = new AddExp(loc, fs.lwr, fs.upr);
if (typeCombine(ea, sc))
    return setError();
fs.prm.type = ea.type;
fs.lwr = ea.e1;
fs.upr = ea.e2;
```
both lower and upper limits are cast to a common type.
So for example this code `static foreach (i; -2..size_t.max) {pragma(msg, i);}` prints `4294967294u` one time.
The lowering is:
```
uint __key50 = 4294967294u;   // ulong 18446744073709551614LU on 64bits
uint __limit51 = 4294967295u;  // ulong 18446744073709551615LU
for (; __key50 < __limit51; __key50 += 1u)
{
uint i = __key50;
return i;
}
```
With your patch this code shows an error using 32bits DMD.
Is this other bug on the Foreach implementation? Should we keep the current behaviour?

Reply:
@ibuclaw Just patched this little problem, as I already have the index type I just need to force the type of the limits:
```
rangefe.lwr.type = indexty;
rangefe.upr.type = indexty;
```
before your code.

Reply:
> @ibuclaw Just patched this little problem, as I already have the index type I just need to force the type of the limits:
> 
> ```
> rangefe.lwr.type = indexty;
> rangefe.upr.type = indexty;
> ```
> 
> before your code.
> 
> I mean this is to keep the current behavior.

It just be OK to use `AddExp/typeCombine` to infer the `indexty` type if `rangefe.lwr.type != rangefe.upr.type`.

Reply:
That is indexty, the index type after the lowering.

Reply:
@ibuclaw, Thanks, the changes are up.